nexus/types/src/deployment/planning_input.rs

```diff
@@ -5,34 +5,136 @@
 //! Types describing inputs the Reconfigurator needs to plan and produce new
 //! blueprints.
 
+use super::AddNetworkResourceError;
+use super::OmicronZoneExternalIp;
+use super::OmicronZoneNetworkResources;
+use super::OmicronZoneNic;
 use crate::external_api::views::PhysicalDiskPolicy;
 use crate::external_api::views::PhysicalDiskState;
 use crate::external_api::views::SledPolicy;
 use crate::external_api::views::SledProvisionPolicy;
 use crate::external_api::views::SledState;
 use clap::ValueEnum;
+use ipnetwork::IpNetwork;
 use omicron_common::address::IpRange;
 use omicron_common::address::Ipv6Subnet;
 use omicron_common::address::SLED_PREFIX;
 use omicron_common::api::external::Generation;
-use omicron_common::api::external::MacAddr;
-use omicron_common::api::internal::shared::SourceNatConfig;
 use omicron_common::api::internal::shared::SourceNatConfigError;
 use omicron_common::disk::DiskIdentity;
-use omicron_uuid_kinds::ExternalIpUuid;
 use omicron_uuid_kinds::OmicronZoneUuid;
 use omicron_uuid_kinds::PhysicalDiskUuid;
 use omicron_uuid_kinds::SledUuid;
 use omicron_uuid_kinds::ZpoolUuid;
-use schemars::JsonSchema;
 use serde::Deserialize;
 use serde::Serialize;
 use std::collections::btree_map::Entry;
 use std::collections::BTreeMap;
-use std::net::IpAddr;
-use std::net::SocketAddr;
 use strum::IntoEnumIterator;
-use uuid::Uuid;
+
+/// Policy and database inputs to the Reconfigurator planner
+///
+/// The primary inputs to the planner are the parent (either a parent blueprint
+/// or an inventory collection) and this structure. This type holds the
+/// fleet-wide policy as well as any additional information fetched from CRDB
+/// that the planner needs to make decisions.
+///
+/// The current policy is pretty limited.  It's aimed primarily at supporting
+/// the add/remove sled use case.
+///
+/// The planning input has some internal invariants that code outside of this
+/// module can rely on. They include:
+///
+/// - Each Omicron zone has at most one external IP and at most one vNIC.
+/// - A given external IP or vNIC is only associated with a single Omicron
+///   zone.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct PlanningInput {
```

Comment:
Decided to move this up-top because this file is called `planning_input.rs`, and generally I think the main part of a struct should live at the top. (Also I was getting really confused :) )

The only differences here are that we now have a `network_resources` field and method.